contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3112,6 +3119,309 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+		 * If cur_value_is_null is true, then the function pg_sequence_last_value
+		 * has returned a NULL value, which means either no rows have been 
+		 * inserted into the table yet, or TRUNCATE TABLE command has been used
+		 * to delete all rows. In this case, after DBCC CHECKIDENT the next
+		 * row inserted will have new_reseed_value as the identity value.
+		 */
+		if (cur_value_is_null)
+		{
+			if (dbcc_stmt.new_reseed_value)
+			{
+				snprintf(message, sizeof(message), "Checking identity information: current identity value 'NULL'.\n");
+
+				DirectFunctionCall3(setval3_oid,
+					ObjectIdGetDatum(seqid),
+					Int64GetDatum(reseed_value),
+					BoolGetDatum(false));
+			}
+			else
+			{
+				snprintf(message, sizeof(message), "Checking identity information: current identity value 'NULL', current column value 'NULL'.\n");
+			}
+		}
+
+		else
+		{
+			if (dbcc_stmt.new_reseed_value)
+			{
+				/* 
+				* Print informational messages if NO_INFOMSGS is not passed as a
+				* DBCC command option.
+				*/
+				if (!dbcc_stmt.no_infomsgs)
+					snprintf(message, sizeof(message), "Checking identity information: current identity value '%ld'.\n", cur_identity_value);
+
+				DirectFunctionCall2(setval_oid,
+					ObjectIdGetDatum(seqid),
+					Int64GetDatum(reseed_value));
+			}
+			else
+			{	
+				SPI_connect();
+				query = psprintf("SELECT MAX(%s) FROM %s.%s", attname,
+								schema_name, table_name);
+				rc = SPI_execute(query, true, 0);
+
+				if (rc != SPI_OK_SELECT)
+					elog(ERROR, "SPI_execute failed: %s", SPI_result_code_string(rc));
+
+				max_identity_value_str = SPI_getvalue(SPI_tuptable->vals[0],
+										SPI_tuptable->tupdesc, 1);
+
+				if(max_identity_value_str)
+					max_identity_value = pg_strtoint64(max_identity_value_str);
+
+				if (!dbcc_stmt.no_infomsgs)
+					snprintf(message, sizeof(message), "Checking identity information: current identity value '%ld', current column value '%s'.\n", 
+														cur_identity_value, 
+														max_identity_value_str ? max_identity_value_str : "NULL");
+
+				pfree(query);
+				SPI_freetuptable(SPI_tuptable);
+
+				/*
+				* RESEED option only resets the identity column value if the 
+				* current identity value for a table is less than the maximum 
+				* identity value stored in the identity column.
+				*/
+				if (dbcc_stmt.is_reseed && max_identity_value_str &&
+					cur_identity_value < max_identity_value)
+				{
+					DirectFunctionCall2(setval_oid,
+						ObjectIdGetDatum(seqid),
+						Int64GetDatum(max_identity_value));
+				}
+			}
+		}
+		
+		if (is_cross_db)
+            SetCurrentRoleId(current_user_id, false);
+	}
+	PG_CATCH();
+	{
+		PG_RE_THROW();
+	    if (is_cross_db)
+            SetCurrentRoleId(current_user_id, false);
+	}
+	PG_END_TRY();
+	
+	if (max_identity_value_str)
+		pfree(max_identity_value_str);
+	pfree(schema_name);
+	pfree(table_name);
+	pfree(nsp_name);
+
+	if(!dbcc_stmt.new_reseed_value)
+	{
+		UnlockRelationOid(table_oid, ShareLock);
```

Comment:
what about exclusive lock?

Reply:
So, we are trying to hold this shareLock on table only for the execution of this statement. We try to hold Exclusive Lock on a transaction level. We want to release this share lock as soon as we are done with the execution of DBCC CHECKIDENT() stmt. But Exclusive Lock will get release when transection gets commited or rollbacked.